File: 백준/Gold/1717. 집합의 표현/집합의 표현.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
using namespace std;

vector<int> arr;
void unionSet(int a,int b);
int find(int num);
int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int n,m;
	cin>>n>>m;

	arr = vector<int>(n+1,0);
	for (int i = 0; i < n + 1; i++) {
		arr[i] = i;
	}

	for (int i = 0; i < m; i++) {
		int input,a,b;
		cin>>input>>a>>b;
		if (input == 0) {
			unionSet(a,b);
		}
		else if (input == 1) {
			int a_jang = find(a);
			int b_jang = find(b);
			if (a_jang == b_jang) {
				cout<<"yes"<<"\n";
			}
			else {
				cout<<"no"<<"\n";
			}
		}
	}

	return 0;
}

void unionSet(int a, int b) {
	a = find(a);
	b = find(b);
	if (a != b) {
		arr[b] = a;
	}
}

int find(int num) {
	if (num == arr[num]) {
		return num;
	}
	else {
		int value = find(arr[num]);
		arr[num] = value;
		return value;
	}
}

File: 백준/Gold/1717. 집합의 표현/집합의 표현.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CodingStudy_CS.Week_14
{
    internal class Program
    {
        static List<int> arr = new List<int>();

        public static void Main(string[] args)
        {
            string[] input = Console.ReadLine().Split(' ');
            int n = int.Parse(input[0]);
            int m = int.Parse(input[1]);

            // 초기화
            for (int i = 0; i <= n; i++)
                arr.Add(i);
            while(m-->0)
            {
                string[] input2 = Console.ReadLine().Split(' ');
                int num = int.Parse(input2[0]);
                int a = int.Parse(input2[1]);
                int b = int.Parse(input2[2]);

                if(num == 0)
                    Union(a, b);
                else
                {
                    int an = Find(a);
                    int bn = Find(b);
                    if(an == bn)
                        Console.WriteLine("YES");
                    else
                        Console.WriteLine("NO");

                }
            }
        }

        static void Union(int a,int b)
        {
            int an = Find(a);
            int bn = Find(b);
            if (an != bn)
                arr[bn] = an;
        }

        static int Find(int num)
        {
            if (num == arr[num])
                return num;
            return arr[num] = Find(arr[num]); ;
        }
    }
}
